src/main/scala/akka/persistence/jdbc/query/JournalSequenceActor.scala

```diff
@@ -25,6 +27,22 @@ object JournalSequenceActor {
   private case object AssumeMaxOrderingIdTimerKey
 
   private type OrderingId = Long
+
+  /**
+    * Efficient representation of missing elements using NumericRanges.
+    * It can be seen as a collection of OrderingIds
+    */
+  private case class MissingElements(elements: Seq[NumericRange[OrderingId]]) {
+    def addRange(from: OrderingId, until: OrderingId): MissingElements = {
+      val newRange = from.until(until)
+      MissingElements(elements :+ newRange)
```

Comment:
Don't we need to check that `newRange` is not already added? Previously, we were writing:

```scala
case e if (previousElement + 1 until e).forall(givenUp.contains) => missing
```
So, when the range was already there, we were just skipping it. The `addRange` could do the same, no?

Reply:
Well, I'm reconsidering that. Actually, I think it's impossible to produce the same range or a sub-range, so such a check will always return false we the range will be added anyway.

Reply:
Previously were were writing `givenUp.contains` however the MissingElements which we are accumulating are not the missing elements which are `givenUp`.

This has actually already been handled in the code of the actor:
```
           // we must decide if we move the cursor forward
            val newMax =
              if ((currentMax + 1).until(currentElement).forall(givenUp.contains)) {
               // we move the cursor forward when:
               // 1) they have been detected as missing on previous iteration, it's time now to give up
               // 2) current + 1 == currentElement (meaning no gap). Note that `forall` on an empty range always returns true
                currentElement
              } else currentMax
  
           // we accumulate in newMissing the gaps we detect on each iteration
           val newMissing =
             if (previousElement + 1 == currentElement || newMax == currentElement) missing
             else missing.addRange(previousElement + 1, currentElement)
```

We compute newMax first which contains the check `((currentMax + 1).until(currentElement).forall(givenUp.contains))`

Because of this we can re-use the information (the value of newMax) in the condition for newMissing: ` if (previousElement + 1 == currentElement || newMax == currentElement) missing`

Reply:
Fair enough. 

I will merge this one then.